packages/react-reconciler/src/__tests__/ReactIncrementalErrorHandling-test.internal.js

```diff
@@ -34,6 +35,115 @@ describe('ReactIncrementalErrorHandling', () => {
     return {type: 'span', children: [], prop};
   }
 
+  xit('recovers from errors asynchronously', () => {
+    class ErrorBoundary extends React.Component {
+      state = {error: null};
+      componentDidCatch(error) {
+        ReactNoop.yield('componentDidCatch');
+        this.setState({error});
+      }
+      render() {
+        if (this.state.error) {
+          ReactNoop.yield('ErrorBoundary (catch)');
+          return <ErrorMessage error={this.state.error} />;
+        }
+        ReactNoop.yield('ErrorBoundary (try)');
+        return this.props.children;
+      }
+    }
+
+    function ErrorMessage(props) {
+      ReactNoop.yield('ErrorMessage');
+      return <span prop={`Caught an error: ${props.error.message}`} />;
+    }
+
+    function Indirection(props) {
+      ReactNoop.yield('Indirection');
+      return props.children;
+    }
+
+    function BadRender() {
+      ReactNoop.yield('throw');
+      throw new Error('oops!');
+    }
+
+    ReactNoop.render(
+      <ErrorBoundary>
+        <Indirection>
+          <Indirection>
+            <Indirection>
+              <BadRender />
+            </Indirection>
+          </Indirection>
+        </Indirection>
+      </ErrorBoundary>,
+    );
+
+    ReactNoop.flushThrough([
+      'ErrorBoundary (try)',
+      'Indirection',
+      'Indirection',
+      'Indirection',
+      // The error was thrown, but React ran out of time and yielded
+      // before recovering.
+      'throw',
+    ]);
+
+    // Upon resuming, componentDidCatch is called
+    ReactNoop.flushThrough(['componentDidCatch', 'ErrorBoundary (catch)']);
+
+    // Flush the rest of the work
+    expect(ReactNoop.flush()).toEqual(['ErrorMessage']);
+    expect(ReactNoop.getChildren()).toEqual([span('Caught an error: oops!')]);
+  });
+
+  it('calls componentDidCatch multiple times for multiple errors', () => {
+    let id = 0;
+    class BadMount extends React.Component {
+      componentDidMount() {
+        throw new Error(`Error ${++id}`);
+      }
+      render() {
+        ReactNoop.yield('BadMount');
+        return null;
+      }
+    }
+
+    class ErrorBoundary extends React.Component {
+      state = {errorCount: 0};
+      componentDidCatch(error) {
+        ReactNoop.yield(`componentDidCatch: ${error.message}`);
+        this.setState(state => ({errorCount: state.errorCount + 1}));
+      }
+      render() {
+        if (this.state.errorCount > 0) {
+          return <span prop={`Number of errors: ${this.state.errorCount}`} />;
+        }
+        ReactNoop.yield('ErrorBoundary');
+        return this.props.children;
+      }
+    }
+
+    ReactNoop.render(
+      <ErrorBoundary>
+        <BadMount />
+        <BadMount />
+        <BadMount />
+      </ErrorBoundary>,
+    );
+
+    expect(ReactNoop.flush()).toEqual([
+      'ErrorBoundary',
+      'BadMount',
+      'BadMount',
+      'BadMount',
+      'componentDidCatch: Error 1',
+      'componentDidCatch: Error 2',
+      'componentDidCatch: Error 3',
```

Comment:
This makes sense theoretically but how practical is it? Once we're in an error state it's likely other things can break due to mistaken assumptions with cryptic errors that can't be reproduced otherwise.

I guess a logger could remember whether it saw an error earlier, and attach that as metadata.

Reply:
I think in most cases, if you catch multiple errors, it will be because they can from different siblings. (Another difference between this implementation and the previous one is that if there's an error, we continue rendering the siblings before we unwind.) In that case, you'd want to log all three errors.

Reply:
Realizing this test made more sense when it was using the render-phase version of the lifecycle. I'll rewrite.